lib/engine/game/g_1849/step/buy_train.rb

```diff
@@ -34,6 +37,34 @@ def can_sell?(entity, bundle)
 
             super
           end
+
+          def buyable_trains(entity)
+            # Cannot buy E-train without E-token
+            trains_to_buy = super
+
+            trains_to_buy = trains_to_buy.reject { |t| t.name == 'E' }
+            trains_to_buy << e_train if @game.electric_dreams? && can_buy_e?(entity)
+            trains_to_buy.uniq
+          end
+
+          def e_train
+            @depot.depot_trains.find { |t| t.name == 'E' }
+          end
+
+          def can_buy_e?(entity)
+            entity.e_token &&
+              e_train.price <= entity.cash
```

Comment:
Can you not e_buy an E train if you have the e token? If you can, then `e_train.price <= entity.cash` isn't needed. `buyable_trains` method will auto handle it. You only need to know if the corp has an e token.